### DISCUSSION TITLE
Multiple password active reset links allowed - security risk?

### DISCUSSION THREAD

**Discussion by jamie-tillman** [keycloak/keycloak#26644] [Q&A]
A recent pentest performed on our Keycloak installation reported that Keycloak allows the user to generate multiple password reset links, all of which will remain valid, even after others are used.  The pentest consultant claims this is a security problem due to a potential brute force attack on the reset token.  Is there any workaround available to prevent or mitigate this, apart from reducing the link expiration to an extremely low timeout?

Here's the comment from the pentest report:

> "Broken Authentication (Multiple Password Reset Links Allowed)"
>
> Steps to reproduce:
> 1. Request to change password for a certain user
> 2. Repeat above step any number of times.
> 3. Go to the user’s registered email’s inbox and click on the reset link in any of the password reset links received due to Step 2 
> 4. Any one of the links will allow password to be reset.
> 
> Impact:
> During the password reset scenario, the reset link (or the reset token part of that link) serves as the credential that will eventually lead to access to the user’s account. As per security principles, multiple valid credentials cannot be allowed for a single account. An attacker could initiate reset links for a user a large number of times and mount a brute force attack to guess the reset token. Trying to guess one value out of hundreds of thousands of correct candidates is far easier than trying to guess only one possible correct value
> Please ensure the following for password reset links:
> 1. Password reset link for a user should only be usable once. (This is currently implemented properly)
> 2. If user requests password reset multiple times, server should allow reset only from the last/latest email link generated for that user. Older links should be invalidated as soon as a new one has been requested by the user.

**Comment by rkkotnik** [comment]
Hi. Did you managed to overcome this scenario somehow?

**Reply by jamie-tillman** [reply]
> Hi. Did you managed to overcome this scenario somehow?

No. We don't have the expertise or time to implement a correction at this time.
